src/rust/cryptography-x509-validation/src/lib.rs

```diff
@@ -10,3 +10,297 @@ pub mod ops;
+};
+use ops::CryptoOps;
+use policy::{Policy, PolicyError};
+use trust_store::Store;
+use types::DNSName;
+
+#[derive(Debug, PartialEq, Eq)]
+pub enum ValidationError {
+    Policy(PolicyError),
+}
+
+impl From<PolicyError> for ValidationError {
+    fn from(value: PolicyError) -> Self {
+        ValidationError::Policy(value)
+    }
+}
+
+#[derive(Default)]
+pub struct AccumulatedNameConstraints<'a> {
+    pub permitted: Vec<GeneralName<'a>>,
+    pub excluded: Vec<GeneralName<'a>>,
+}
+
+pub type Chain<'c> = Vec<Certificate<'c>>;
+type IntermediateChain<'c> = (Chain<'c>, AccumulatedNameConstraints<'c>);
+
+pub fn verify<'leaf: 'chain, 'inter: 'chain, 'store: 'chain, 'chain, B: CryptoOps>(
+    leaf: &'chain Certificate<'leaf>,
+    intermediates: impl IntoIterator<Item = Certificate<'inter>>,
+    policy: &Policy<'_, B>,
+    store: &'chain Store<'store>,
+) -> Result<Chain<'chain>, ValidationError> {
+    let builder = ChainBuilder::new(HashSet::from_iter(intermediates), policy, store);
+
+    builder.build_chain(leaf)
+}
+
+struct ChainBuilder<'a, 'inter, 'store, B: CryptoOps> {
+    intermediates: HashSet<Certificate<'inter>>,
+    policy: &'a Policy<'a, B>,
+    store: &'a Store<'store>,
+}
+
+// When applying a name constraint, we need to distinguish between a few different scenarios:
+// * `Applied(true)`: The name constraint is the same type as the SAN and matches.
+// * `Applied(false)`: The name constraint is the same type as the SAN and does not match.
+// * `Skipped`: The name constraint is a different type to the SAN.
+enum ApplyNameConstraintStatus {
+    Applied(bool),
+    Skipped,
+}
+
+impl ApplyNameConstraintStatus {
+    fn is_applied(&self) -> bool {
+        matches!(self, Applied(_))
+    }
+
+    fn is_match(&self) -> bool {
+        match self {
+            Applied(a) => *a,
+            _ => false,
+        }
+    }
+}
+
+impl<'a, 'inter, 'store, 'leaf, 'chain, 'work, B: CryptoOps> ChainBuilder<'a, 'inter, 'store, B>
+where
+    'leaf: 'chain,
+    'inter: 'chain,
+    'store: 'chain,
+    'work: 'leaf + 'inter,
+    'chain: 'work,
+{
+    fn new(
+        intermediates: HashSet<Certificate<'inter>>,
+        policy: &'a Policy<'a, B>,
+        store: &'a Store<'store>,
+    ) -> Self {
+        Self {
+            intermediates,
+            policy,
+            store,
+        }
+    }
+
+    fn potential_issuers(
+        &'a self,
+        cert: &'a Certificate<'work>,
+    ) -> impl Iterator<Item = &'a Certificate<'work>> + '_ {
+        // TODO: Optimizations:
+        // * Use a backing structure that allows us to search by name
+        //   rather than doing a linear scan
+        // * Search by AKI and other identifiers?
+        self.intermediates
+            .iter()
+            // NOTE: The intermediate set isn't allowed to offer a self-signed
+            // certificate as a candidate, since self-signed certs can only
+            // be roots.
+            .filter(|&candidate| *candidate != *cert)
```

Comment:
On second look, me neither 🙂 -- I think we might have added this before the `cert_is_self_signed` helper, which performs the documented check here.

CC @tetsuo-cpp, in case you remember.

Reply:
I'm not sure why we chose do to this instead of checking with `cert_is_self_signed`. We did this because there was a Limbo case that triggered a hang. I'll check whether this still passes with `cert_is_self_signed` and if so, let's use that instead.

Reply:
Ok, done. I've opted to use `cert_is_self_issued` instead of `cert_is_self_signed` as I don't think we want to check the key here (that's expensive and we already do it elsewhere).

[d498f67](https://github.com/pyca/cryptography/pull/8873/commits/d498f67c19a97778aa9739b0b69fb9edeed49d0d)

Reply:
Ok done. I had to make `ops` public on the `Policy` (I believe it was that way previously). If we'd rather not do that, I can easily make a wrapper function on `Policy` that calls into `cert_is_self_signed`.

[c40761f](https://github.com/pyca/cryptography/pull/8873/commits/c40761f209f097d89d3d4e1570b89fe3cc317979)